### PR TITLE
Add custom job intervals

### PR DESCRIPTION
**Pull request by yoonthegoon**
# Closes: #1226
# What's Changed

Adds custom cron style interval for running jobs.
image

# TODO

- [x] Explanation of Change(s)
- [x] Attached Screenshots, Payload Example
- [x] Unit, Integration Tests
- [x] Documentation Updates (when adding/changing features)

### CODE REVIEW

File: nautobot/extras/models/jobs.py [hunk 1]
```diff
@@ -762,6 +762,13 @@ class ScheduledJob(BaseModel):
         verbose_name="Approval date/time",
         help_text="Datetime that the schedule was approved",
     )
+    custom = models.CharField(
```
Comment (jathanism) [inline nautobot/extras/models/jobs.py:765]: This name needs to be more descriptive. I would prefer that it be called `custom_schedule` or even just `crontab` so that it's clear that it if it is set, it will be used in that way. On its own `custom` is not quite clear I don't think.

File: nautobot/extras/models/jobs.py [hunk 2]
```diff
@@ -762,6 +762,13 @@ class ScheduledJob(BaseModel):
...
+        max_length=255,
+        blank=True,
+        null=True,
```
Comment (jathanism) [inline nautobot/extras/models/jobs.py:768]: We shouldn't ever use `null=True` on `CharField`. (I know there are existing examples of this, but I'd like to not see it propagate. Thanks!)

```suggestion
        blank=True,
```

**Review by jathanism — CHANGES_REQUESTED** [review]
Looking good, thanks! Just need some clarifications.

**Review by yoonthegoon — COMMENTED** [review]
all changes have been made

File: nautobot/extras/api/views.py [hunk 3]
```diff
@@ -344,6 +344,8 @@ def _create_schedule(serializer, data, commit, job, job_model, request):
     else:
         time = serializer["start_time"]
         name = serializer["name"]
+    crontab = serializer.get("crontab")
+    crontab = "" if crontab is None else crontab
```
Comment (jtdub) [inline nautobot/extras/api/views.py:348]: `crontab = serializer.get("crontab", "")`

Remove 348

**Review by jtdub — CHANGES_REQUESTED** [review]

**Comment by yoonthegoon** [comment]
integration test passes on my fork

File: nautobot/extras/migrations/0036_auto_20220721_1356.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
+# Generated by Django 3.2.14 on 2022-07-21 18:56
```
Comment (jathanism) [inline nautobot/extras/migrations/0036_auto_20220721_1356.py:1]: Please delete these migrations and recreate them so that there is only a single migration file.

File: nautobot/extras/forms.py [hunk 5]
```diff
@@ -1047,6 +1054,24 @@ def clean(self):
...
+                    ).split(" ")
+                    schedules.crontab(
+                        minute=minute,
+                        hour=hour,
+                        day_of_month=day_of_month,
+                        month_of_year=month_of_year,
+                        day_of_week=day_of_week,
+                    )
+                except Exception as e:
+                    raise ValidationError({"_recurrence_custom_time": e})
```
Comment (jathanism) [inline nautobot/extras/forms.py:1073]: This appears to be duplicating the same code as is found in `ScheduledJob.get_crontab()`. Can we not just use that same method? There's nothing on that method that requires an instance, so it could be converted to a `class method` and then called here directly from the model class object a la `ScheduledJob.get_crontab()`.

File: nautobot/extras/models/jobs.py [hunk 6]
```diff
@@ -804,6 +810,9 @@ def clean(self):
         # bitwise xor also works on booleans, but not on complex values
         if bool(self.approved_by_user) ^ bool(self.approved_at):
             raise ValidationError("Approval by user and approval time must either both be set or both be undefined")
+        if self.interval == JobExecutionType.TYPE_CUSTOM:
+            if self.get_crontab() is str:
+                raise ValidationError(self.get_crontab())
```
Comment (jathanism) [inline nautobot/extras/models/jobs.py:815]: This pattern seems weird to be evaluating the return value of the method using an identity check (`is str`) and then calling the same method again to raise its return value as an exception. If it's expected that `get_crontab()` should return an error, it would be best to just let that error occur and expect the callers to handle it.

It also appears that you're not actually doing anything with the result of `get_crontab()` in the event that it doesn't raise an error.

So IMO if we had `get_crontab()` just raise a `ValidationError` then you'd only have to call `get_crontab()` here to assert that the crontab is accurate or raises an error.

File: nautobot/extras/models/jobs.py [hunk 7]
```diff
@@ -817,6 +826,32 @@ def schedule(self):
...
+            minute, hour, day_of_month, month_of_year, day_of_week = self.crontab.split(" ")
+            return schedules.crontab(
+                minute=minute,
+                hour=hour,
+                day_of_month=day_of_month,
+                month_of_year=month_of_year,
+                day_of_week=day_of_week,
+            )
+        except Exception as e:
+            return e
```
Comment (jathanism) [inline nautobot/extras/models/jobs.py:853]: IMO just return this call without exception handling, or ALWAYS have it raise a `ValidationError` exception. Simply returning an exception is not a common pattern and I think it can lead to unexpected results. Such as `below where you `return self.crontab()`

Another thing to consider is that in the original source from `django-celery-beat`, the `CrontabSchedule` model is also attempting to use TZ-aware crontabs. Please, see: https://github.com/celery/django-celery-beat/blob/master/django_celery_beat/models.py#L324-L342

File: nautobot/extras/views.py [hunk 8]
```diff
@@ -1024,6 +1028,7 @@ def post(self, request, class_path=None, slug=None):
                     one_off=schedule_type == JobExecutionType.TYPE_FUTURE,
                     user=request.user,
                     approval_required=job_model.approval_required,
+                    crontab=crontab,
```
Comment (jathanism) [inline nautobot/extras/views.py:1031]: Now that I'm looking at this again, it appears that there may be a couple ways to do custom-Crontab based schedules incorrectly. Because what if you set `interval=JobExecutionType.TYPE_CUSTOM, one_off=True, but then set `crontab=crontab`. It doesn't appear that we have any checks to assert that these combinations of options are mutualy exclusive.

I would like to see some more defensiveness inside of the `ScheduledJob.clean()` method here and also have these possible combinations tested to make sure that we are thinking about the edge cases and protecting against them with proper error-handling and sanity checks!

File: nautobot/extras/models/jobs.py [hunk 9]
```diff
@@ -825,6 +860,8 @@ def to_cron(self):
             return schedules.crontab(minute=t.minute, hour=t.hour)
         elif self.interval == JobExecutionType.TYPE_WEEKLY:
             return schedules.crontab(minute=t.minute, hour=t.hour, day_of_week=t.weekday())
+        elif self.interval == JobExecutionType.TYPE_CUSTOM:
+            return self.get_crontab()
```
Comment (jathanism) [inline nautobot/extras/models/jobs.py:864]: Just to be pedantic, in the current implementation if `crontab` is malformed, this would return the value of an exception, and not a properly-parsed crontab object, which I do not think is what we would actually want.

**Review by jathanism — CHANGES_REQUESTED** [review]
Thanks for the updates! I provided some more feedback.

File: nautobot/extras/views.py (hunk 8 above)
Comment (yoonthegoon) [reply nautobot/extras/views.py:1031]: Hmm, one both `extras/views.py` and `extras/api/views.py`, `one_off` doesn't seem to be set by the user. In either case, it's only ever set to true when the schedule type is equal to `JobExecutionType.TYPE_FUTURE`
I'll see about another defense in the `clean()` method after resolving everything else and testing.

File: nautobot/extras/models/jobs.py (hunk 9 above)
Comment (yoonthegoon) [reply nautobot/extras/models/jobs.py:864]: yeah, changing the get_crontab() method to only return a crontab() would fix this

File: nautobot/extras/models/jobs.py (hunk 6 above)
Comment (yoonthegoon) [reply nautobot/extras/models/jobs.py:815]: removing this while also removing the try except in the get_crontab() method should resolve

File: nautobot/docs/additional-features/job-scheduling-and-approvals.md [hunk 10]
```diff
@@ -29,6 +31,10 @@ http://nautobot/api/extras/jobs/$JOB_ID/run/ \
 --data '{"schedule": {"name": "test", "interval": "future", "start_time": "2030-01-01T01:00:00.000Z"}}'
 ```
 
+For custom interval, a `crontab` parameter must be added.
+
+`--data '{"schedule": {"name": "test", "interval": "custom", "start_time": "2030-01-01T01:00:00.000Z", "crontab": "*/15 * * * *"}}'`
```
Comment (jathanism) [inline nautobot/docs/additional-features/job-scheduling-and-approvals.md:36]: This seems like a bit of a mixed case here. If there's a `crontab` why do we need to provide a `start_time`? Wouldn't the crontab determine the start time? Perhaps similar to "immediately" start times, perhaps we should consider using `datetime.now()` internally (See: `nautobot.extras.api.views._create_schedule()`)?

File: nautobot/extras/tests/test_views.py [hunk 11]
```diff
@@ -881,6 +881,8 @@ def test_only_enabled_is_listed(self):
         self.assertHttpStatus(response, 200)
         self.assertNotIn("test4", extract_page_body(response.content.decode(response.charset)))
 
+    # TODO: add `def test_non_valid_crontab_syntax(self):`
```
Comment (jathanism) [inline nautobot/extras/tests/test_views.py:884]: This is looking great. I would like to see some tests here as we had indicated as what happens when `ScheduledJob.get_crontab()` goes wrong.

And then also asserting that we can correctly create Crontab-based scheduled jobs via the API as well as the UI.

**Review by jathanism — CHANGES_REQUESTED** [review]
Looking great. Let's just get this tested and wrap it up, assuming my new comments aren't too much.

File: nautobot/docs/additional-features/job-scheduling-and-approvals.md (hunk 10 above)
Comment (bryanculver) [reply nautobot/docs/additional-features/job-scheduling-and-approvals.md:36]: They both should exist.

> Starting `start_time = time() + 1 week`, run this task every 15 minutes.

Comment (yoonthegoon) [reply nautobot/docs/additional-features/job-scheduling-and-approvals.md:36]: i can have `start_time` be optional. when not specified, start immediately. i'm imagining cases where you'd want a job to run every 15 minutes, but you put that together on a friday and want it to start monday morning or something.

Comment (lampwins) [reply nautobot/docs/additional-features/job-scheduling-and-approvals.md:36]: Yeah, this is in line with the existing functionality where start time can act as a deferral on the start of a continual schedule.

Comment (jathanism) [reply nautobot/docs/additional-features/job-scheduling-and-approvals.md:36]: > i can have `start_time` be optional. when not specified, start immediately. i'm imagining cases where you'd want a job to run every 15 minutes, but you put that together on a friday and want it to start monday morning or something.

That's fair. In that case, please follow yours and Bryan's wisdom and disregard what we discussed in private.

File: nautobot/extras/models/jobs.py [hunk 12]
```diff
@@ -825,6 +860,8 @@ def to_cron(self):
             return schedules.crontab(minute=t.minute, hour=t.hour)
         elif self.interval == JobExecutionType.TYPE_WEEKLY:
             return schedules.crontab(minute=t.minute, hour=t.hour, day_of_week=t.weekday())
+        elif self.interval == JobExecutionType.TYPE_CUSTOM:
+            return self.get_crontab(self.crontab)
         raise ValueError(f"I do not know to convert {self.interval} to a Cronjob!")
```
Comment (bryanculver) [inline nautobot/extras/models/jobs.py:865]: I think it would be good to add a housekeeping item to stash the crontab onto `self.crontab` in all cases. Could clean up a lot of these methods.

**Review by bryanculver — APPROVED** [review]
LGTM

File: nautobot/extras/api/nested_serializers.py [hunk 13]
```diff
@@ -173,13 +173,24 @@ def validate(self, data):
...
                 raise serializers.ValidationError(
                     {
                         "start_time": "Please enter a valid date and time greater than or equal to the current date and time."
                     }
                 )
 
+            if data["interval"] == choices.JobExecutionType.TYPE_CUSTOM:
+
+                if data.get("crontab") is None:
+                    raise serializers.ValidationError({"crontab": "ya need sth"})
```
Comment (jathanism) [inline nautobot/extras/api/nested_serializers.py:188]: "ya need sth"? What is that! 😁

File: nautobot/extras/api/views.py [hunk 14]
```diff
@@ -343,7 +344,7 @@ def _create_schedule(serializer, data, commit, job, job_model, request):
     elif type_ == JobExecutionType.TYPE_CUSTOM:
         time = serializer.get("start_time")  # doing .get("key", "default") returns None instead of "default"
         if time is None:
-            time = timezone.now()
+            time = timezone.now() + timedelta(seconds=20)
```
Comment (jathanism) [inline nautobot/extras/api/views.py:347]: I seem to recall you mentioning previously why you added 20 seconds, but would you please mind adding a brief comment here for posterity?

File: nautobot/extras/tests/test_views.py [hunk 15]
```diff
@@ -881,7 +882,93 @@ def test_only_enabled_is_listed(self):
...
+            ScheduledJob.objects.create(
+                enabled=True,
+                name="test10",
+                task="nautobot.extras.jobs.scheduled_job_handler",
+                job_class="local/test_pass/TestPass",
+                interval=JobExecutionType.TYPE_CUSTOM,
+                user=self.user,
+                start_time=timezone.now(),
+                crontab=None,
+            )
```
Comment (jathanism) [inline nautobot/extras/tests/test_views.py:898]: Since this pattern is repeated multiple times with seemingly only the argument to `crontab` changing, can you please move this into a factory method that can be called taking a `crontab` parameter?

**Review by jathanism — CHANGES_REQUESTED** [review]
Thank you for adding the tests! I just have a couple minor asks and we're good to wrap this up!

File: nautobot/extras/api/nested_serializers.py (hunk 13 above)
Comment (yoonthegoon) [reply nautobot/extras/api/nested_serializers.py:188]: LOL i put TODOs on stupid comments and debugs so they wouldn't commit but i guess i missed one

**Review by jathanism — APPROVED** [review]
*chef's kiss* 👨‍🍳 💋 Thank you for your contribution! This is great!

**Comment by yoonthegoon** [comment]
thank you for the feedback. i learned a lot
